src/ophyd_async/epics/adcore/_core_writer.py

```diff
@@ -0,0 +1,219 @@
+from ophyd_async.core._providers import DatasetDescriber, NameProvider, PathProvider
+from ophyd_async.core._signal import (
+    observe_value,
+    set_and_wait_for_value,
+    wait_for_value,
+)
+from ophyd_async.core._status import AsyncStatus
+from ophyd_async.core._utils import DEFAULT_TIMEOUT
+
+# from ophyd_async.epics.adcore._core_logic import ADBaseDatasetDescriber
+from ._core_io import (
+    ADBaseDatasetDescriber,
+    Callback,
+    NDArrayBaseIO,
+    NDFileIO,
+    NDPluginBaseIO,
+)
+from ._utils import FileWriteMode
+
+NDFileIOT = TypeVar("NDFileIOT", bound=NDFileIO)
+ADWriterT = TypeVar("ADWriterT", bound="ADWriter")
+
+
+class ADWriter(DetectorWriter, Generic[NDFileIOT]):
+    default_suffix: str = "FILE1:"
+
+    def __init__(
+        self,
+        fileio: NDFileIOT,
+        path_provider: PathProvider,
+        name_provider: NameProvider,
+        dataset_describer: DatasetDescriber,
+        file_extension: str = "",
+        mimetype: str = "",
+        plugins: dict[str, NDPluginBaseIO] | None = None,
+    ) -> None:
+        self._plugins = plugins or {}
+        self.fileio = fileio
+        self._path_provider = path_provider
+        self._name_provider = name_provider
+        self._dataset_describer = dataset_describer
+        self._file_extension = file_extension
+        self._mimetype = mimetype
+        self._last_emitted = 0
+        self._emitted_resource = None
+
+        self._capture_status: AsyncStatus | None = None
+        self._multiplier = 1
+        self._filename_template = "%s%s_%6.6d"
+
+    @classmethod
+    def with_io(
+        cls: type[ADWriterT],
+        prefix: str,
+        path_provider: PathProvider,
+        dataset_source: NDArrayBaseIO | None = None,
+        fileio_suffix: str | None = None,
+        plugins: dict[str, NDPluginBaseIO] | None = None,
+    ) -> ADWriterT:
+        try:
+            fileio_cls = get_args(cls.__orig_bases__[0])[0]  # type: ignore
+        except IndexError as err:
+            raise RuntimeError("File IO class for writer not specified!") from err
+
+        fileio = fileio_cls(prefix + (fileio_suffix or cls.default_suffix))
+        dataset_describer = ADBaseDatasetDescriber(dataset_source or fileio)
+
+        def name_provider() -> str:
+            if fileio.parent == "Not attached to a detector":
+                raise RuntimeError("Initializing writer without parent detector!")
+            return fileio.parent.name
+
+        writer = cls(
+            fileio, path_provider, name_provider, dataset_describer, plugins=plugins
+        )
+        return writer
+
+    async def begin_capture(self) -> None:
+        info = self._path_provider(device_name=self._name_provider())
+
+        await self.fileio.enable_callbacks.set(Callback.ENABLE)
+
+        # Set the directory creation depth first, since dir creation callback happens
+        # when directory path PV is processed.
+        await self.fileio.create_directory.set(info.create_dir_depth)
```

Comment:
When does the directory actually get created? I am building off of this branch for #576 and am running into an issue in the tests where the following fails:
https://github.com/bluesky/ophyd-async/pull/606/files#diff-e9e5dca7c989f7d7b0226116ef60f0c95d525f3905412a6c5993ef25afcff02cR115-R117

Reply:
The directories are created by the areaDetector file plugin on a callback that is run when you write to the file path signal. In that callback it gets the value of `CreateDirectory` and uses it as the number of directories it is permitted to set. There is a `DirectoryExists` PV that then says if the target directory exists (whether or not any were created).  If you have access to your SimDetector IOC from the other week, you can test this by setting up the tiff plugin to write to `/tmp` for example. 

In the tests, since the IOC behavior is not available, we need to add callbacks to signals manually to replicate the behavior.

See the factory fixture for generating AD ophyd objects for an example: https://github.com/jwlodek/ophyd-async/blob/a616fd22e656202a1f7c65fbf4afeb2adaea5dba/tests/epics/conftest.py#L51